### PR TITLE
*: Provide staleReadProcessor to process stale read

### LINKED ISSUE #32697: Provide staleReadProcessor to process stale read

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #32697

### What is changed and how it works?

This is one of the PRs to refactor stale read. The full draft is here: #32325

This modifications of the PR includes:

- Introduce a new package `staleread` to provide stale read related functions.
- Introduce a new interface `staleread.Processor` to provide methods processing and fetching stale read context. 
- Move some of the stale read processing codes to `staleread.staleReadProcessor` which implements `staleread.Processor` to simplify `Preprocessor` . 

For the PRs in the future, we'll do more things refactoring stale read, including:

- Move more stale read codes (prepare/execute/begin) to processor.
- Separate the processor the some sub classes for different scenes (Prepare, CreateView, InitTxnContext ...).
- Provide `StaleReadTxnContextProvider` and set it to `TxnManager` in processor

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/preprocess.go [hunk 1]
```diff
@@ -184,6 +188,8 @@ type preprocessor struct {
 	tableAliasInJoin []map[string]interface{}
 	withName         map[string]interface{}
 
+	staleReadProcessor staleread.Processor
```
Comment (JmPotato) [inline planner/core/preprocess.go:191]: Though I'm OK to make this change, does this mean that we may have more and more `xxxProcessor` fields in the future? Do we have any idea on how to simplify this part?

File: planner/core/preprocess.go [hunk 2]
```diff
@@ -1620,117 +1626,27 @@ func (p *preprocessor) handleAsOfAndReadTS(node *ast.AsOfClause) {
...
+	if p.IsStaleness = p.staleReadProcessor.IsStaleness(); p.IsStaleness {
+		p.LastSnapshotTS = p.staleReadProcessor.GetStalenessReadTS()
+		p.SnapshotTSEvaluator = p.staleReadProcessor.GetStalenessTSEvaluatorForPrepare()
+		p.InfoSchema = p.staleReadProcessor.GetStalenessInfoSchema()
+	}
+
+	if p.IsStaleness || p.ctx.GetSessionVars().GetReplicaRead().IsClosestRead() {
+		p.ReadReplicaScope = config.GetTxnScopeFromConfig()
+	} else {
+		p.ReadReplicaScope = p.ctx.GetSessionVars().TxnCtx.TxnScope
```
Comment (JmPotato) [inline planner/core/preprocess.go:1647]: IMO a comment is needed here to explain the Local Txn compatibility.

Comment (lcwangchao) [reply planner/core/preprocess.go:1647]: Okay, I'll add it later

File: planner/core/preprocess.go (hunk 1 above)
Comment (lcwangchao) [reply planner/core/preprocess.go:191]: Having a `staleReadProcessor` here is to decouple the stale read code from preprocessor and expect to reuse it. It is only a choice for stale read and I think for other scenes the design will not be limited to 'xxxProcessor'.

For stale read, in the future, there is still only one processor to process stale read. However, we can have multiple subclasses inherits `staleread.Processor` and we can use one of them according to the exact scenario

File: sessiontxn/staleread/processor.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessiontxn"
+	"github.com/pingcap/tidb/table/temptable"
+)
+
+// StalenessTSEvaluator is a function to get staleness ts
+type StalenessTSEvaluator func(sctx sessionctx.Context) (uint64, error)
+
+// Processor is an interface used to process stale read
+type Processor interface {
```
Comment (xhebox) [inline sessiontxn/staleread/processor.go:33]: Consider `var _ Proccessor = &staleReadProcessor{}` to enforce the implementation?

**Review by xhebox — COMMENTED** [review]
LGTM overall

File: sessiontxn/staleread/processor.go (hunk 3 above)
Comment (lcwangchao) [reply sessiontxn/staleread/processor.go:33]: fine. Good idea

**Review by xhebox — APPROVED** [review]

File: planner/core/preprocess.go [hunk 4]
```diff
@@ -1620,117 +1626,32 @@ func (p *preprocessor) handleAsOfAndReadTS(node *ast.AsOfClause) {
...
+
+	if p.IsStaleness = p.staleReadProcessor.IsStaleness(); p.IsStaleness {
+		p.LastSnapshotTS = p.staleReadProcessor.GetStalenessReadTS()
+		p.SnapshotTSEvaluator = p.staleReadProcessor.GetStalenessTSEvaluatorForPrepare()
+		p.InfoSchema = p.staleReadProcessor.GetStalenessInfoSchema()
+	}
+
+	// It is a little hacking for the below codes. `ReadReplicaScope` is used both by stale read's closest read and local txn.
+	// They are different features and the value for `ReadReplicaScope` will be conflicted in some scenes.
+	// But because local txn is still an experimental feature, we should make stale read work first.
```
Comment (Yisaer) [inline planner/core/preprocess.go:1646]: In which cases that `ReadReplicaScope` will get conflict?

Comment (lcwangchao) [reply planner/core/preprocess.go:1646]: I think it is in local txn. It is used by local txn to check all the tables in a query must be located in the zone 
 where `ReadReplicaScope` value is (https://github.com/pingcap/tidb/blob/74d74b0a7b6bee275d392199bb319859682627ff/executor/point_get.go#L470-L479). But `set @@tidb_replica_read = 'closest-replicas'` will also set `ReadReplicaScope` , that will cause  the local txn check fails

Comment (Yisaer) [reply planner/core/preprocess.go:1646]: I think the reason is that the behavior of the local transaction is not well-defined yet, so we can't define the behavior between `local txn` and `tidb_replica_read `. This should recorded as an issue.

File: sessiontxn/staleread/processor.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+// NewStaleReadProcessor creates a new stale read processor
+func NewStaleReadProcessor(sctx sessionctx.Context) Processor {
+	p := &staleReadProcessor{}
+	p.init(sctx)
+	return p
+}
+
+// OnSelectTable will be called when process table in select statement
+func (p *staleReadProcessor) OnSelectTable(tn *ast.TableName) error {
```
Comment (Yisaer) [inline sessiontxn/staleread/processor.go:140]: Why the comment for explaining the code logic removed?

File: sessiontxn/staleread/main_test.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	"go.uber.org/goleak"
+)
+
+func TestMain(m *testing.M) {
+	opts := []goleak.Option{
+		goleak.IgnoreTopFunction("go.opencensus.io/stats/view.(*worker).start"),
+	}
+	testbridge.SetupForCommonTest()
+	goleak.VerifyTestMain(m, opts...)
+}
```
Comment (Yisaer) [inline sessiontxn/staleread/main_test.go:30]: Can this moved into `processor_test` instead of creating `main_test`?

**Review by Yisaer — COMMENTED** [review]
rest lgtm

File: planner/core/preprocess.go (hunk 4 above)
Comment (JmPotato) [reply planner/core/preprocess.go:1646]: Here is the issue: https://github.com/pingcap/tidb/issues/31772.

Comment (lcwangchao) [reply planner/core/preprocess.go:1646]: There is an open issue here: https://github.com/pingcap/tidb/issues/31772 . I think the best way to resolve it is to introduce a new state like `TxnScope` for local txn. But local txn is not GA yet, we can do it in other PRs later.

File: sessiontxn/staleread/main_test.go (hunk 6 above)
Comment (lcwangchao) [reply sessiontxn/staleread/main_test.go:30]: fine

File: sessiontxn/staleread/processor.go (hunk 5 above)
Comment (lcwangchao) [reply sessiontxn/staleread/processor.go:140]: I'll complete it later

Comment (lcwangchao) [reply sessiontxn/staleread/processor.go:140]: done

**Review by Yisaer — APPROVED** [review]

**Pull request merged**